Added ability to disable accordion controls.

jellisii opened the pull request:
Occasionally I want to lock an accordion into it's current state.  This allows the disabled attribute to be put on an accordion container, disabling the toggle, down, and up events, and styling the cursor appropriately when the accordion is disabled.

jellisii committed:
Added ability to disable accordion controls.

IamManchanda commented on docs/pages/accordion.md:
````diff
@@ -125,6 +125,39 @@ By default, at least one pane in an accordion must be open. This can be changed
+
+There may be times where you want to disable pane switching on an accordion. This can be accomplished by setting the disabled attribute.
+
+```html
````
`html_example` would be better!

jellisii replied:
I used the rest of the file as the basis for the modification.  No where in the rest of the file is `html_example` used.  If this needs to be changed, I can do that.

jellisii committed:
Altered all ```html tags to be ```html_example per suggestion from @IamManchanda.

jellisii commented:
It would figure I'd push this change when AWS has an outage... :D

IamManchanda commented:
Yeah @jellisi its the hot topic right now :D

jellisii commented:
Can I get a feel for if this will get accepted?  I'm working on some things that this particular functionality makes SIGNIFICANTLY easier to deal with...

denisahac commented:
@IamManchanda Line number 132 of **accordion.md** file denotes the start of Markdown code block and specify what language to use for syntax highlighting, in this case HTML.

@jellisii please read this section of [Markdown Code and Syntax Highligthing](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#code-and-syntax-highlighting)

That denotes what language syntax to use in the code block, in that case the HTML `html_example` isn't a valid language. And you've taken a wrong information from @IamManchanda

jellisii committed:
Fixed docs per @denisahac

jellisii commented:
There seem to be inconsistent application of `html` and `html_example` in other places in the docs.  There probably should be something set, to quote Douglas Adams, "if not straight, permanently bent" with regard to this.

IamManchanda commented:
I am on travel for next 1.5 hours... Give me 2 hour and I will give a brief about this 

html_example is valid as far as foundation docs is concerned!!!

IamManchanda commented:
Ohk so as promissed here's my brief 

As far as foundation docs ( style sherpa ) is concerned, 

`html` lets you create a code block in Markdown, whereas
`html_example` lets you not only create a code block but also lets you show a live demo of the component below the code sample

Source: http://foundation.zurb.com/sites/docs/style-sherpa.html#code-samples

----------------------------------------------------------------------------------------------------------

> please read this section of [Markdown Code and Syntax Highligthing](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#code-and-syntax-highlighting)

That markdown cheatsheet differ's from Zurb's style sherpa ! 

----------------------------------------------------------------------------------------------------------------

@jellisii @denisahac .... when i said `html_example` would be better, i meant it should be better to show the demo of the code also for disable accordion's .... My advice would be to use `html_example` where ever demo needs to be shown for the code.....

IamManchanda commented on docs/pages/accordion.md:
````diff
@@ -139,7 +172,7 @@ Add the attribute `data-deep-link="true"` to a accordion to:
 - allow users to open a particular accordion panel at page load with a hash-appended URL
 
-```html_example
+```html
````
Please revert this back,
Demo needs to be shown for this!

IamManchanda commented on docs/pages/accordion.md:
````diff
@@ -165,6 +198,6 @@ For example, <a target="_blank" href="#deeplink3">http://example.com/#deeplink3<
 When linking directly to a accordion panel, it might not be obvious that the content appears within a accordion panel. An additional attribute `data-deep-link-smudge` rolls the page up slightly after deep linking (to a horizontal accordion) so that the accordion is at the top of the viewport.
 
-```html_example
+```html
````
Please revert this back,
Demo needs to be shown for this!

jellisii replied:
So style sherpa is authoritative for Foundation documentation?  Asking for clarification, as I can alter my own submission to reflect it if that's the case.

IamManchanda replied:
> So style sherpa is authoritative for Foundation documentation

Yes ofcourse @jellisii .... Style sherpa and Panini is behind Foundation docs !

IamManchanda reviewed (changes requested):
As per brief i gave, i think you need to revert these two back for sure and even the disable accordion one

jellisii committed:
Re-factored to reflect style-sherpa docs

Per @IamManchanda.

jellisii commented:
This should be square.

jellisii commented:
Any word on this getting considered for 6.3.2, since the 6.3.1 ship has sailed?

kball commented on js/foundation.accordion.js:
```diff
@@ -153,6 +153,9 @@ class Accordion {
    * @function
    */
   toggle($target) {
+    if ($target.closest('.accordion').is('[disabled]')) {
```
Since we're determining behavior, this should probably look for the data attribute not the class (we use classes for display/styles, data attributes for behavior/js)... so something like
```
if ($target.closest('[data-accordion]').is('[disabled]')) {
```

kball commented on js/foundation.accordion.js:
```diff
@@ -178,6 +181,13 @@ class Accordion {
+     * checking firstTime allows for initial render of the accordion
+     * to render preset is-active panes.
+     */
+    if ($target.closest('.accordion').is('[disabled]') && !firstTime)  {
```
Same as above, should target data attribute

kball commented:
This is new functionality, so will probably put it for 6.4 rather than 6.3.2, but would love to get the above notes addressed and merge this for that, due out next month.

jellisii committed:
Merge remote-tracking branch 'upstream/develop' into develop

jellisii commented:
So in working on this, I discovered a edge case that I may want to refactor, but I'd like some opinions on it from a YetiNaut.

When the `disabled` attribute is set on an accordion, it disables all of the up, down, and toggle methods for said accordion.  This could potentially cause heartburn for a developer if they try to manipulate a disabled accordion like in a heavily managed UI.  I can address this in one of two ways.  

* Least impact lift is document the need to remove the `disabled` attribute from the accordion being manipulated.
* Potential ease-of-use/better developer experience is alter the JavaScript to accept a second argument to the `up` and `toggle` methods that behaves like `firstTime` does on `down`:
```javascript
down($target, firstTime) {
    /**
     * checking firstTime allows for initial render of the accordion
     * to render preset is-active panes.
     */
    if ($target.closest('[data-accordion]').is('[disabled]') && !firstTime)  {
      console.info('Cannot call down on an accordion that is disabled.');
      return;
    }
...
```

@kball, could I get some input on this?

jellisii committed:
Changes to better dovetail with current Foundation design patterns.

Changed method of detecting disabled per
https://github.com/zurb/foundation-sites/pull/9812#discussion_r109806300

Altered docs to reflect current functionality of disable accordions,
documented below.

If an accordion is disabled, all motion methods (`up`, `down`, `toggle`)
get eaten, with the exception of `down` when combined with `firstTime`.
Added a `console.info()` message to indicate when these methods are
attempted on a disabled accordion.

jellisii commented:
I went ahead and went with the least impact lift, adding some `console.warn` statements to help the developer when troubleshooting issues around it.

kball commented:
@jellisii sorry about the slow response, was heads down on another project... this is looking good to me, are you happy with it and feel it's ready to merge?

jellisii commented:
Yeah, I think it's done for now.  Thank you for your time @kball!

kball commented:
All right, merged! Thanks @jellisii, and congrats on your first Foundation contribution!
